### DISCUSSION TITLE
Override @Substitute

### DISCUSSION THREAD

**Discussion by fsudrot** [quarkusio/quarkus#22613] [Q&A] [answered]
For production purpose, I'd like to use mongodb+srv connection style. But Quarkus does not allow such a style due to substitutions we found in io.quarkus.mongodb.runtime.graal.MongoClientSubstitutions, like this one :

```java
@Substitute
public List<String> resolveHostFromSrvRecords(final String srvHost) {
    throw new UnsupportedOperationException("mongo+srv:// not supported in native mode");
}
```

If I try to provide my own substitution to allow a custom dns resolution, I get an error :
```
Error: Already registered: com.mongodb.internal.dns.DefaultDnsResolver.resolveHostFromSrvRecords(String)
```
Is there a mean to "override" the quarkus provided substitution to provide custom ones ?

**Comment by gsmet** [comment, marked as answer]
IIRC, we disable it because it doesn't work in native :). @cescoffier would know more (btw, Clément, maybe you could add a Javadoc on the substitution explaining why it's disabled? I had a look at the history and I couldn't find any concrete info).

That being said, it's possible to override substitutions by removing them from the jar and provide new ones. You have an example here: https://github.com/quarkusio/quarkus/blob/main/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java#L121 . It requires an extension but extensions are really easy to create with the Maven command scaffolding.

Another option is to use `quarkus.class-loading.removed-resources."groupId:artifactId"=your/package/YourClass.class` but this class loading config is a bit special so maybe experiment with it and if you can get things working, I would recommend the extension approach.

**Comment by cescoffier** [comment]
The reason for this substitution is because `mongo+srv://` requires JNDI which is not supported in native. See https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/main/com/mongodb/internal/dns/DefaultDnsResolver.java. 

I have contemplated the idea to implement my own DNS lookup, but it would behave differently in native and in JVM. So, the safest was to just disable that support. 

There is now the possibility to implement your own resolver (so my custom resolver would be used in JVM and native), but I didn't try that yet.

**Reply by cescoffier** [reply]
As a follow-up, here is the issue in native if you remove the substitution:
```
2022-01-05 13:58:32,515 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: com.mongodb.MongoClientException: Unable to support mongodb+srv// style connections as the 'com.sun.jndi.dns.DnsContextFactory' class is not available in this JRE. A JNDI context is required for resolving SRV records.
```

So, we would need a custom DNS resolver that does not use JNDI.

**Reply by fsudrot** [reply]
Thanks to the suggestions.
I have applied the second from @gsmet.
I first add 
```yaml
quarkus:
  class-loading:
    removed-resources:
      "io.quarkus:quarkus-mongodb-client": io/quarkus/mongodb/runtime/graal/DefaultDnsResolverSubstitution.class
```
Then I provide my own substitution class using dnsjava to make SRV and TXT resolution.
dnsjava is not directly ready for native application so I have to add ```--initialize-at-run-time=org.xbill.DNS```
But it works, I finally have mongodb+srv working in native mode.
Thanks to all quarkus team, that's a great product !!!

**Reply by cescoffier** [reply]
An alternative (which is the one I would pursue) is to use the vertx dns client (native friendly, fast and lightweight, already in the classpath)

Be careful that with your approach you have a behavior difference between JVM and native. You can prevent it by passing your custom resolver.

**Reply by marcogutto** [reply]
Hi @fsudrot !
I'm facing the same issue here.

Just to clarify you add the 

```
quarkus:
  class-loading:
    removed-resources:
      "io.quarkus:quarkus-mongodb-client": io/quarkus/mongodb/runtime/graal/DefaultDnsResolverSubstitution.class
```

on application.properties and mapped in your pom the org.xbill.DNS?

Where you configured the `--initialize-at-run-time=org.xbill.DNS`?

@cescoffier How can I setup the vertx dns client in this case?

@gsmet thanks for this solution.

Thanks in advance guys!

**Reply by fsudrot** [reply]
On @cescoffier advice, I switch to vertx dnsClient. It's still working on native mode in production. That was a big feature for the ops.

And to answer @marcogutto, it was an application.yml, that's because the format is indented but the key is the same for application.properties : quarkus.class-loading.removed-resources."io.quarkus\:quarkus-mongodb-client"
as explained [here](https://quarkus.io/guides/class-loading-reference#hidingremoving-classes-and-resources-from-dependencies)

For the --initialize-at-runtime, that's also in application.properties : 
quarkus.native.additional-build-args=--initialize-at-run-time=org.xbill.DNS
as explained [here](https://quarkus.io/guides/writing-native-applications-tips#delaying-class-initialization)

And to use vertx, I just create a dnsclient :
```java
static DnsClient createDnsClient() {
        List<String> dnsServers = nameServers();
        return Vertx.vertx().createDnsClient(new DnsClientOptions()
                .setPort(53)
                .setHost(dnsServers.get(0))
                .setQueryTimeout(10000)
        );
    }
 ```
 and use it to resolve srv and txt in the substituted method
 ```java
 Future<List<SrvRecord>> resolveSRV = dnsClient.resolveSRV("_mongodb._tcp." + srvHost);
 
 Future<List<String>> resolveTXT = dnsClient.resolveTXT(srvHost);
```

not very fancy but it works

**Reply by gsmet** [reply]
@cescoffier I wonder if we should implement it even if a bit hackyish and open issues in the MongoDB client tracker to allow for the customization of the resolver and clean it up later.

**Reply by cescoffier** [reply]
@gsmet I agree, we need to provide something. However, my previous attempt failed... 

@fsudrot, there are a few issues with your code.
You should use something like:

```java
static DnsClient client; // Cache the client

    static synchronized DnsClient getDnsClient() {
        if (client == null) {
            Vertx vertx = Arc.container().instance(Vertx.class).get(); // Use the managed instance, so the client is closed and use the right event loops
            int port = ConfigProvider.getConfig().getOptionalValue("mongo.dns.port", Integer.class).orElse(53); // Extract configuration
            String server = ConfigProvider.getConfig().getValue("mongo.dns.server", String.class);
            client = vertx.createDnsClient(port, server);
        }
        return client;
    }

    static synchronized String resolve() {
        Duration timeout = ConfigProvider.getConfig().getOptionalValue("mongo.dns.lookup-timeout", Duration.class).orElse(Duration.ofSeconds(5));
        List<SrvRecord> records = getDnsClient().resolveSRV("...").await().atMost(timeout); // Mutiny API, so we get the await construct and check that we can block the current thread.
        List<String> txt = getDnsClient().resolveTXT("...").await().atMost(timeout);
        //...
    }
```

**Reply by fsudrot** [reply]
@cescoffier thanks for the correction, always functionnal in production

**Reply by marcogutto** [reply]
Hi @fsudrot !
Thanks for the feedback!

I implemented a sample [here](https://github.com/marcogutto/quarkus-reactive-mongodb.git) and it did not worked yet ... 
There are something that I'm missing.

Can you help me ?

Thanks in advance!

**Reply by fsudrot** [reply]
@marcogutto 
You evict the quarkus DefaultDnsResolverSubstitution
```
quarkus.class-loading.removed-resources."io.quarkus\:quarkus-mongodb-client"=io/quarkus/mongodb/runtime/graal/DefaultDnsResolverSubstitution.class
```
But I do not see any custom implementation for DefaultDnsResolver, have I missed something ?

**Reply by fsudrot** [reply]
@marcogutto 
I add two classes to add the ability to resolve SRV and TXT to the mongo client in :
https://github.com/fsudrot/quarkus-reactive-mongodb/tree/feature/mongodb_srv
I do not test but it is certianly more functional like this

**Reply by cescoffier** [reply]
I just opened https://github.com/quarkusio/quarkus/pull/22944.

**Comment by gsmet** [comment]
@cescoffier added support for it for 2.7: https://github.com/quarkusio/quarkus/pull/22944

@fsudrot it would be nice if you could try it once CR1 is released (should be tomorrow).

**Reply by fsudrot** [reply]
I'll check the CR1 arrival

**Reply by fsudrot** [reply]
@gsmet It seems to work properly in 2.7.0.CR1
Thanks

**Reply by cescoffier** [reply]
Thanks for the feedback @fsudrot

**Reply by marcogutto** [reply]
Hi Guys!

I'm testing the 2.7.0 and get the error when build in native mode:

Error: Already registered: com.mongodb.internal.dns.DefaultDnsResolver.sameParentDomain(List, String)
#15 188.3 com.oracle.svm.core.util.UserError$UserException: Already registered: com.mongodb.internal.dns.DefaultDnsResolver.sameParentDomain(List, String)

I added in my application.properties:
quarkus.mongodb.native.dns.use-vertx-dns-resolver=true
quarkus.mongodb.native.dns.lookup-timeout=10s

Are there more configuration?

Thanks in advance!

**Reply by cescoffier** [reply]
Can you provide a reproducer? Maybe you have the substitutions twice.

**Reply by marcogutto** [reply]
Sorry, I forgot the older class package.

Now, it builds normally but I get the error:
(vert.x-eventloop-thread-1) Unable to look up TXT record for host serverlessinstance0.oagmp.mongodb.net



Thanks in advance!

**Reply by marcogutto** [reply]
In my contaiver resolv.conf file:

bash-4.4$ cat /etc/resolv.conf 
nameserver 127.0.0.11
options ndots:0

**Reply by cescoffier** [reply]
Pretty sure it uses ndots: 1 by default. Any reason to use 0?

**Reply by marcogutto** [reply]
I really don't know.

I'm using the default native image quay.io/quarkus/quarkus-micro-image:1.0.

I think that it comes with this default value.

**Reply by cescoffier** [reply]
Hum... Ok, I will need to reproduce it. Can you open an issue so I don't forget?

**Reply by marcogutto** [reply]
Hi @cescoffier I opened this issue https://github.com/quarkusio/quarkus/issues/23548.

Thanks in advance!
